Implement browser.storageArea.getKeys() for Web Extension Storage API.

MoeBazziGIT opened the pull request:
#### 972a6ce2e9ed3c6b75eedc37b7d53112c036fef1
<pre>
Implement browser.storageArea.getKeys() for Web Extension Storage API.
<a href="https://bugs.webkit.org/show_bug.cgi?id=280275">https://bugs.webkit.org/show_bug.cgi?id=280275</a>

Reviewed by Timothy Hatcher.

This patch implements a web extension API to retrieve all keys for a given storage area in the
browser.storage API.

WECG Proposal: <a href="https://github.com/w3c/webextensions/blob/main/proposals/storage-get-keys.md">https://github.com/w3c/webextensions/blob/main/proposals/storage-get-keys.md</a>
WECG Original Issue: w3c/webextensions#601

This patch also includes the following clean-ups:
- Change &apos;auto&apos; to &apos;auto *&apos; when assigning the result of WebExtensionContext::storageForType()
 to explicitly denote pointer type.
- Update strongSelf retrieval to in _WKWebExtensionStorageSQLiteStore.mm methods to leverage
 ARC in extension code, allowing the use of Objective-C pointers instead of RetainPtr, which
is only necessary in other parts of WebKit that do not utilize ARC.
- Remove unnecessary explicit template types from WeakObjCPtr in _WKWebExtensionStorageSQLiteStore.

* Source/WebKit/UIProcess/Extensions/Cocoa/API/WebExtensionContextAPIStorageCocoa.mm:
(WebKit::WebExtensionContext::storageGet):
(WebKit::WebExtensionContext::storageGetKeys):
(WebKit::WebExtensionContext::storageGetBytesInUse):
* Source/WebKit/UIProcess/Extensions/Cocoa/_WKWebExtensionStorageSQLiteStore.h:
* Source/WebKit/UIProcess/Extensions/Cocoa/_WKWebExtensionStorageSQLiteStore.mm:
(-[_WKWebExtensionStorageSQLiteStore getAllKeys:]):
(-[_WKWebExtensionStorageSQLiteStore getValuesForKeys:completionHandler:]):
(-[_WKWebExtensionStorageSQLiteStore getStorageSizeForKeys:completionHandler:]):
(-[_WKWebExtensionStorageSQLiteStore getStorageSizeForAllKeysIncludingKeyedData:withCompletionHandler:]):
(-[_WKWebExtensionStorageSQLiteStore setKeyedData:completionHandler:]):
(-[_WKWebExtensionStorageSQLiteStore deleteValuesForKeys:completionHandler:]):
* Source/WebKit/UIProcess/Extensions/WebExtensionContext.h:
* Source/WebKit/UIProcess/Extensions/WebExtensionContext.messages.in:
* Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPIStorageAreaCocoa.mm:
(WebKit::WebExtensionAPIStorageArea::getKeys):
* Source/WebKit/WebProcess/Extensions/API/WebExtensionAPIStorageArea.h:
* Source/WebKit/WebProcess/Extensions/Interfaces/WebExtensionAPIStorageArea.idl:
* Tools/TestWebKitAPI/Tests/WebKitCocoa/WKWebExtensionAPIStorage.mm:
(TestWebKitAPI::TEST(WKWebExtensionAPIStorage, Errors)):
(TestWebKitAPI::TEST(WKWebExtensionAPIStorage, GetKeys)):

Canonical link: <a href="https://commits.webkit.org/284602@main">https://commits.webkit.org/284602@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/b4568afafb22b89dc59d7412757d64c79470628c

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/69939 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/49340 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/22692 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/74024 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/21097 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/72056 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/57140 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/20948 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/55502 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/13979 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/73005 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/44949 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/60321 "Passed tests") | [❌ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/35982 "Found 1 new API test failure: /WPE/TestWebKitWebContext:/webkit/WebKitWebContext/languages (failure)") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/41614 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/17749 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/19474 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/63537 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/18100 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/75739 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/87/builds/14164 "Built successfully") | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/17329 "Build is in progress. Recent messages:OS: Sonoma (14.5), Xcode: 15.4; Running apply-patch; Checked out pull request; Running run-layout-tests-in-stress-mode; 17 flakes") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/63197 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/86/builds/14200 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/60396 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/63132 "Passed tests") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/88/builds/11152 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/4759 "Passed tests") | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/10681 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/45143 "Built successfully") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/46217 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/47488 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/45958 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

xeenon commented on Source/WebKit/UIProcess/Extensions/Cocoa/_WKWebExtensionStorageSQLiteStore.mm:
```diff
@@ -71,14 +71,34 @@ - (instancetype)initWithUniqueIdentifier:(NSString *)uniqueIdentifier storageTyp
+        NSSet<NSString *> *keysSet = [self _getAllKeysReturningErrorMessage:&errorMessage];
+
+        dispatch_async(dispatch_get_main_queue(), ^{
+            completionHandler(keysSet.allObjects, errorMessage);
```
```suggestion
        auto *keysArray = [self _getAllKeysReturningErrorMessage:&errorMessage].allObjects;

        dispatch_async(dispatch_get_main_queue(), ^{
            completionHandler(keysArray, errorMessage);
```

MoeBazziGIT replied:
✅

xeenon commented on Source/WebKit/UIProcess/Extensions/Cocoa/_WKWebExtensionStorageSQLiteStore.mm:
```diff
@@ -71,14 +71,34 @@ - (instancetype)initWithUniqueIdentifier:(NSString *)uniqueIdentifier storageTyp
+{
+    auto weakSelf = WeakObjCPtr<_WKWebExtensionStorageSQLiteStore> { self };
+    dispatch_async(_databaseQueue, ^{
+        auto strongSelf = weakSelf.get();
```
Not new to this, but the pointer should be noted here (and other places in this file.)

```suggestion
        auto *strongSelf = weakSelf.get();
```

MoeBazziGIT replied:
Ah okay, so the type of `strongSelf` (`RetainPtr`) has implicit conversion operators to the underlying pointer?

xeenon replied:
Ah I forgot `weakSelf.get()` returned a `RetainPtr`. Just keep this as plain `auto` still then.

xeenon replied:
Or this needs an extra `get()`:

```suggestion
        auto *strongSelf = weakSelf.get().get();
```

MoeBazziGIT replied:
Depends, do you think having the pointer notation is worth having a `.get().get()`? I might update it across other files as well.

xeenon replied:
If it is in extension code, I think ObjC pointer is better, since we use ARC in our files. Otherwise `RetainPtr` is needed for all other uses in WebKit since they don't use ARC.

MoeBazziGIT replied:
Made an in-PR commit for this ✅

xeenon commented on Source/WebKit/UIProcess/Extensions/Cocoa/_WKWebExtensionStorageSQLiteStore.mm:
```diff
@@ -71,14 +71,34 @@ - (instancetype)initWithUniqueIdentifier:(NSString *)uniqueIdentifier storageTyp
 
+- (void)getAllKeys:(void (^)(NSArray<NSString *> *keys, NSString *errorMessage))completionHandler
+{
+    auto weakSelf = WeakObjCPtr<_WKWebExtensionStorageSQLiteStore> { self };
```
This syntax should work now, I fixed `WeakObjCPtr` recently to support it.

```suggestion
    auto weakSelf = WeakObjCPtr { self };
```

MoeBazziGIT replied:
Sounds good. I've gone ahead and updated all the instances across the codebase I could find that could use this. What's the policy on cleanup patches? ...should I make a new PR for that or add the new commit to this one?

xeenon replied:
If it is just in this file, you can do it in this PR. Or if it is broader than just extension files, just file a new bug and post a clean up PR.

MoeBazziGIT replied:
Made an in-PR commit for just the `_WKWebExtensionStorageSQLiteStore.mm` file. Will make another PR for the changes to the rest of the codebase, including other web extension code.

Is there a standard way of posting a cleanup PR relative to this PR?

xeenon commented on Source/WebKit/UIProcess/Extensions/Cocoa/API/WebExtensionContextAPIStorageCocoa.mm:
```diff
@@ -64,6 +64,19 @@
+{
+    auto callingAPIName = makeString("browser.storage."_s, toAPIString(dataType), ".getKeys()"_s);
+
+    auto storage = storageForType(dataType);
```
Should be a pointer too.
```suggestion
    auto *storage = storageForType(dataType);
```

MoeBazziGIT replied:
Sounds good. I'll update the few others in the file as well.

MoeBazziGIT replied:
Made an in-PR commit for this for all instances in the `_WKWebExtensionStorageSQLiteStore.mm` file

xeenon reviewed (approved):
You should adjust the bug and title of the commit to be `getKeys()` instead of `getAllKeys()`. Also mention the [WECG issue](https://github.com/w3c/webextensions/issues/601) so it cross-links, and hard wrap the commit message to ~110 chars.

Thanks for doing this @MoeBazziGIT!

MoeBazziGIT committed:
Implement browser.storageArea.getKeys() for Web Extension Storage API.
https://bugs.webkit.org/show_bug.cgi?id=280275

Reviewed by Timothy Hatcher.

This patch implements a web extension API to retrieve all keys for a given storage area in the
browser.storage API.

WECG Proposal: https://github.com/w3c/webextensions/blob/main/proposals/storage-get-keys.md
WECG Original Issue: w3c/webextensions#601

This patch also includes the following clean-ups:
- Change 'auto' to 'auto *' when assigning the result of WebExtensionContext::storageForType()
 to explicitly denote pointer type.
- Update strongSelf retrieval to in _WKWebExtensionStorageSQLiteStore.mm methods to leverage
 ARC in extension code, allowing the use of Objective-C pointers instead of RetainPtr, which
is only necessary in other parts of WebKit that do not utilize ARC.
- Remove unnecessary explicit template types from WeakObjCPtr in _WKWebExtensionStorageSQLiteStore.

* Source/WebKit/UIProcess/Extensions/Cocoa/API/WebExtensionContextAPIStorageCocoa.mm:
(WebKit::WebExtensionContext::storageGet):
(WebKit::WebExtensionContext::storageGetKeys):
(WebKit::WebExtensionContext::storageGetBytesInUse):
* Source/WebKit/UIProcess/Extensions/Cocoa/_WKWebExtensionStorageSQLiteStore.h:
* Source/WebKit/UIProcess/Extensions/Cocoa/_WKWebExtensionStorageSQLiteStore.mm:
(-[_WKWebExtensionStorageSQLiteStore getAllKeys:]):
(-[_WKWebExtensionStorageSQLiteStore getValuesForKeys:completionHandler:]):
(-[_WKWebExtensionStorageSQLiteStore getStorageSizeForKeys:completionHandler:]):
(-[_WKWebExtensionStorageSQLiteStore getStorageSizeForAllKeysIncludingKeyedData:withCompletionHandler:]):
(-[_WKWebExtensionStorageSQLiteStore setKeyedData:completionHandler:]):
(-[_WKWebExtensionStorageSQLiteStore deleteValuesForKeys:completionHandler:]):
* Source/WebKit/UIProcess/Extensions/WebExtensionContext.h:
* Source/WebKit/UIProcess/Extensions/WebExtensionContext.messages.in:
* Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPIStorageAreaCocoa.mm:
(WebKit::WebExtensionAPIStorageArea::getKeys):
* Source/WebKit/WebProcess/Extensions/API/WebExtensionAPIStorageArea.h:
* Source/WebKit/WebProcess/Extensions/Interfaces/WebExtensionAPIStorageArea.idl:
* Tools/TestWebKitAPI/Tests/WebKitCocoa/WKWebExtensionAPIStorage.mm:
(TestWebKitAPI::TEST(WKWebExtensionAPIStorage, Errors)):
(TestWebKitAPI::TEST(WKWebExtensionAPIStorage, GetKeys)):

Canonical link: https://commits.webkit.org/284602@main
